### ISSUE TITLE
No Switches to enable/disable IPv6 port forwards

### ISSUE THREAD

**Issue by DavidFW1960** [mammuth/ha-fritzbox-tools#104]
I only see this:
![image](https://user-images.githubusercontent.com/36390762/86069266-c0d4a200-babc-11ea-9361-37ffebd41a82.png)
No switches.
I have a system where I only forward an IPv6 address... IPv4 is not exposed

**Comment by mammuth** [comment]
Any errors in the log? 
I'm not sure whether I saw IPv6 port forwards working at all in the past. At least I'm not using IPv6 forwards.

**Comment by DavidFW1960** [comment]
Just removed and re-added
```
Log Details (ERROR)
Logger: aiohttp.server
Source: custom_components/fritzbox_tools/__init__.py:167
First occurred: 15:38:27 (1 occurrences)
Last logged: 15:38:27

Error handling request
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/urllib3/connection.py", line 160, in _new_conn
    (self._dns_host, self.port), self.timeout, **extra_kw
  File "/usr/local/lib/python3.7/site-packages/urllib3/util/connection.py", line 84, in create_connection
    raise err
  File "/usr/local/lib/python3.7/site-packages/urllib3/util/connection.py", line 74, in create_connection
    sock.connect(sa)
ConnectionRefusedError: [Errno 111] Connection refused

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/urllib3/connectionpool.py", line 677, in urlopen
    chunked=chunked,
  File "/usr/local/lib/python3.7/site-packages/urllib3/connectionpool.py", line 392, in _make_request
    conn.request(method, url, **httplib_request_kw)
  File "/usr/local/lib/python3.7/http/client.py", line 1252, in request
    self._send_request(method, url, body, headers, encode_chunked)
  File "/usr/local/lib/python3.7/http/client.py", line 1298, in _send_request
    self.endheaders(body, encode_chunked=encode_chunked)
  File "/usr/local/lib/python3.7/http/client.py", line 1247, in endheaders
    self._send_output(message_body, encode_chunked=encode_chunked)
  File "/usr/local/lib/python3.7/http/client.py", line 1026, in _send_output
    self.send(msg)
  File "/usr/local/lib/python3.7/http/client.py", line 966, in send
    self.connect()
  File "/usr/local/lib/python3.7/site-packages/urllib3/connection.py", line 187, in connect
    conn = self._new_conn()
  File "/usr/local/lib/python3.7/site-packages/urllib3/connection.py", line 172, in _new_conn
    self, "Failed to establish a new connection: %s" % e
urllib3.exceptions.NewConnectionError: <urllib3.connection.HTTPConnection object at 0x7f172d3920d0>: Failed to establish a new connection: [Errno 111] Connection refused

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/requests/adapters.py", line 449, in send
    timeout=timeout
  File "/usr/local/lib/python3.7/site-packages/urllib3/connectionpool.py", line 725, in urlopen
    method, url, error=e, _pool=self, _stacktrace=sys.exc_info()[2]
  File "/usr/local/lib/python3.7/site-packages/urllib3/util/retry.py", line 439, in increment
    raise MaxRetryError(_pool, url, error or ResponseError(cause))
urllib3.exceptions.MaxRetryError: HTTPConnectionPool(host='10.90.11.100', port=49000): Max retries exceeded with url: /igddesc.xml (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7f172d3920d0>: Failed to establish a new connection: [Errno 111] Connection refused'))

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/aiohttp/web_protocol.py", line 418, in start
    resp = await task
  File "/usr/local/lib/python3.7/site-packages/aiohttp/web_app.py", line 458, in _handle
    resp = await handler(request)
  File "/usr/local/lib/python3.7/site-packages/aiohttp/web_middlewares.py", line 119, in impl
    return await handler(request)
  File "/usr/src/homeassistant/homeassistant/components/http/real_ip.py", line 39, in real_ip_middleware
    return await handler(request)
  File "/usr/src/homeassistant/homeassistant/components/http/ban.py", line 73, in ban_middleware
    return await handler(request)
  File "/usr/src/homeassistant/homeassistant/components/http/auth.py", line 127, in auth_middleware
    return await handler(request)
  File "/usr/src/homeassistant/homeassistant/components/http/view.py", line 129, in handle
    result = await result
  File "/usr/src/homeassistant/homeassistant/components/config/config_entries.py", line 145, in post
    return await super().post(request, flow_id)
  File "/usr/src/homeassistant/homeassistant/components/http/data_validator.py", line 60, in wrapper
    result = await method(view, request, *args, **kwargs)
  File "/usr/src/homeassistant/homeassistant/helpers/data_entry_flow.py", line 106, in post
    result = await self._flow_mgr.async_configure(flow_id, data)
  File "/usr/src/homeassistant/homeassistant/data_entry_flow.py", line 153, in async_configure
    result = await self._async_handle_step(flow, cur_step["step_id"], user_input)
  File "/usr/src/homeassistant/homeassistant/data_entry_flow.py", line 201, in _async_handle_step
    result: Dict = await getattr(flow, method)(user_input)
  File "/config/custom_components/fritzbox_tools/config_flow.py", line 116, in async_step_start_config
    self.fritz_tools = await self.hass.async_add_executor_job(lambda: FritzBoxTools(
  File "/usr/local/lib/python3.7/concurrent/futures/thread.py", line 57, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/config/custom_components/fritzbox_tools/config_flow.py", line 123, in <lambda>
    device_list=[]
  File "/config/custom_components/fritzbox_tools/__init__.py", line 167, in __init__
    address=host, port=port, user=username, password=password, timeout=30.0
  File "/usr/local/lib/python3.7/site-packages/fritzconnection/core/fritzconnection.py", line 109, in __init__
    self.device_manager.add_description(source)
  File "/usr/local/lib/python3.7/site-packages/fritzconnection/core/devices.py", line 61, in add_description
    root = get_xml_root(source, timeout=self.timeout, session=self.session)
  File "/usr/local/lib/python3.7/site-packages/fritzconnection/core/utils.py", line 56, in get_xml_root
    source = get_content_from(source, timeout=timeout, session=session)
  File "/usr/local/lib/python3.7/site-packages/fritzconnection/core/utils.py", line 38, in get_content_from
    with session.get(url) as response:
  File "/usr/local/lib/python3.7/site-packages/requests/sessions.py", line 543, in get
    return self.request('GET', url, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/requests/sessions.py", line 530, in request
    resp = self.send(prep, **send_kwargs)
  File "/usr/local/lib/python3.7/site-packages/requests/sessions.py", line 643, in send
    r = adapter.send(request, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/requests/adapters.py", line 516, in send
    raise ConnectionError(e, request=request)
requests.exceptions.ConnectionError: HTTPConnectionPool(host='10.90.11.100', port=49000): Max retries exceeded with url: /igddesc.xml (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7f172d3920d0>: Failed to establish a new connection: [Errno 111] Connection refused'))
```
I put the wrong IP address in when I re-added it... (router is on .1 not .100)
I do have 2 HA instances... main one is on .100 but I am adding this to a secondary one at .153 (running dev)

**Comment by DavidFW1960** [comment]
ugh... all those errors seem to be from when I added it
Just doing a do-over and there does not seem to be anything in the logs.

I have been using IPv6 for ages with HA (Via Caddy reverse proxy - the domain doesn't even have an A record - that's on my main machine)

**Comment by mammuth** [comment]
Maybe our current code doesn't work with IPv6. 

If you are familiar with python, you could debug it by simply instantiating `fritzconnection` in the REPL and requesting the port forwards as in the referenced method

**Comment by DavidFW1960** [comment]
I'll try the debug log tomorrow (on iPad now) The other suggestion is a foreign language lol.

**Comment by DavidFW1960** [comment]
OK so once I opened up IPv4 it presents the switches however, it only controls IPv4 port forward not ipv6 so unless I create an A record I can't use this unfortunately.

**Comment by mammuth** [comment]
Alright, then we probably need adjustments for IPv6 switches.

**Comment by AaronDavidSchneider** [comment]
Sorry for the late answer. The implementation for the port forwarding only considers IPv4 ports. As stated in the readme:
```
Only works if you have a dedicated IPv4 address (it won't work with DS-Lite)
```
When I implemented the port forwarding switch, I also tried to use `fritzconnection` for IPv6, but it didnt really work.

**Comment by DavidFW1960** [comment]
I don't have DS-Lite I have a real IPv4 and IPv6 address. I don't mind forwarding both as long as the switch disables both.

**Comment by AaronDavidSchneider** [comment]
I understand. The switch however only controls the IPv4 part.

**Comment by DavidFW1960** [comment]
Understood but that doesn't help me

**Comment by AaronDavidSchneider** [comment]
feel free to contribute 😄

**Comment by DavidFW1960** [comment]
Thank for that but that's not helpful. Maybe you could take up amateur brain surgery...

**Comment by mammuth** [comment]
> Thank for that but that's not helpful. Maybe you could take up amateur brain surgery...

Well, everyone here is contributing to this project in their free time while (probably) pursuing full-time jobs, hobbies, having a family etc. 
If some features are missing, someone needs to build it. And this mostly means, someone who is affected by the missing feature, because they have higher intrinsic motivation for developing it.

I understand that it's sad that you cannot control IPv6 port forwards, but you cannot expect that any of the contributors just jumps and develops your feature wish (especially since there are probably many other feature requests which might have even higher impact for all users).

**Comment by DavidFW1960** [comment]
So I reported this as an issue - I didn't know if it should work or if I was doing something wrong or if there was some limitation I was not aware of.

I completely understand that People do this in their spare time out of the goodness of their heart and/or their own needs and they freely provide this to everyone as-is. I don't expect people to jump to it to satisfy my requests.

However it is disrespectful and pisses me off no end when someone suggests I should implement it myself. I don't have the skill set to do that - not even remotely the skill set. It is just rude to make that kind of suggestion. If I could do it myself I would and would be happy to contribute. Unfortunately that exact kind of bullshit response is not uncommon in the HA project. I do maintain a custom component and lovelace card that I have hacked and people have sometimes asked for an additional feature in the custom component and I don't need to make that response to anyone when I decline (usually because I don't have the skills to do it)
